### ISSUE TITLE
Corruption of image files during migration or optimization

### ISSUE THREAD

**Issue by arturvader** [gulpjs/gulp#2803]
When the script is triggered, the file from the app/images/brand_logo.jpg folder is copied to dist/images/brand_logo.jpg. There should be no changes to the file but it changes and becomes unreadable.

In another project, when using "gulp-tinify": "^1.0.2" the same problem.

I checked with gulp version 4.0.2 installed and everything works correctly.

Node version:
20.13.0
npm version:
10.5.2
gulp --version
CLI version: 3.0.0
Local version: 5.0.0

package.json:
{
  "name": "test",
  "version": "1.0.0",
  "main": "gulpfile.js",
  "private": true,
  "type": "module",
  "browserslist": [
    "last 2 version",
    "not dead",
    "not ie <= 11",
    "iOS >= 12"
  ],
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "gulp": "^5.0.0",
    "gulp-imagemin": "^9.1.0",
    "gulp-newer": "^1.4.0"
  }
}

gulpfile.js:
import gulp from 'gulp';
import imagemin from 'gulp-imagemin';
import newer from 'gulp-newer';

const	app = 'app',
		dist = 'dist';

const config = {
	app : {
		svg : app + '/images/**/*.svg',
		img : app + '/images/**/*.{jpg,jpeg,png,gif,webp}',
	},
	dist : {
		svg : dist + '/images/**/*.svg',
		img : dist + '/images/**/*',
		all : dist + '/**/*',
	},
}

// Images

export const images = gulp.series(
	function(cb){
	return gulp.src(config.app.svg, { base: 'app' })
		.pipe(newer(config.dist.svg))
		.pipe(imagemin())
		.pipe(gulp.dest(dist))
		cb();
	},
);
export const imgtinify = () => {
	return gulp.src(config.app.img, { base: 'app', encoding: false, buffer: false })
		.pipe(newer(config.dist.img))
		.pipe(gulp.dest(dist))
};

// Watch

export const startwatch = () => {
	gulp.watch(config.app.img, gulp.series(
		imgtinify,
	));
	gulp.watch(config.app.svg, gulp.series(images, ));
};

// Default

export default gulp.series(
	gulp.parallel(
		images,
		imgtinify,
	),
	gulp.parallel(
		startwatch,
	),
);
<img width="1473" alt="Screenshot 2024-05-08 at 20 43 30" src="https://github.com/gulpjs/gulp/assets/51017968/a8903dba-61c4-403a-bdd8-4b0bf265782c">

**Comment by yocontra** [comment]
Add `{encoding: false}` to any gulp.src call where you are using binary files - see now docs and release blog for v5 for info.

**Comment by arturvader** [comment]
> Add `{encoding: false}` to any gulp.src call where you are using binary files - see now docs and release blog for v5 for info.

Pay attention to the code that I provided as a sample. What you wrote is indicated in the code ( gulp.src(config.app.img, { base: 'app', encoding: false, buffer: false }) ) and it does not help.

**Comment by yocontra** [comment]
Sorry, the code wasn't formatted in the issue very well and I was looking at the SVG logic. With encoding: false it shouldn't make any changes to the file at all.

Just going to repost your code simplified and formatted for reference:


```js
import gulp from 'gulp'
import imagemin from 'gulp-imagemin'
import newer from 'gulp-newer'

const app = 'app'
const dist = 'dist'

const config = {
  app: {
    svg: app + '/images//*.svg',
    img: app + '/images//.{jpg,jpeg,png,gif,webp}'
  },
  dist: {
    svg: dist + '/images/**/.svg',
    img: dist + '/images//*',
    all: dist + '//*'
  }
}
export const imgtinify = () => {
  return gulp
    .src(config.app.img, {
      base: 'app',
      encoding: false,
      buffer: false
    })
    .pipe(newer(config.dist.img))
    .pipe(gulp.dest(dist))
}

export default gulp.series(imgtinify)
```

Are you able to reproduce this with the above code, and does removing gulp-newer also reproduce it?

**Comment by arturvader** [comment]
> Sorry, the code wasn't formatted in the issue very well and I was looking at the SVG logic. With encoding: false it shouldn't make any changes to the file at all.
> 
> Just going to repost your code simplified and formatted for reference:
> 
> ```js
> import gulp from 'gulp'
> import newer from 'gulp-newer'
> 
> const app = 'app'
> const dist = 'dist'
> 
> const config = {
>   app: {
>     img: app + '/images//.{jpg,jpeg,png,gif,webp}'
>   },
>   dist: {
>     img: dist + '/images//*'
>   }
> }
> export const imgtinify = () =>
>   gulp
>     .src(config.app.img, {
>       base: 'app',
>       encoding: false,
>       buffer: false
>     })
>     .pipe(newer(config.dist.img))
>     .pipe(gulp.dest(dist))
> 
> export default gulp.series(imgtinify)
> ```
> 
> Are you able to reproduce this with the above code, and does removing gulp-newer also reproduce it? Could you provide an example of a before and after file so I can look at the binary?

Original image:
![brand_logo](https://github.com/gulpjs/gulp/assets/51017968/d55a677e-84a9-4efd-ba76-88389675ff09)


Resulting file:
![brand_logo](https://github.com/gulpjs/gulp/assets/51017968/c9caeb5d-795e-4ba2-9874-91931d40c39a)

**Comment by arturvader** [comment]
> Sorry, the code wasn't formatted in the issue very well and I was looking at the SVG logic. With encoding: false it shouldn't make any changes to the file at all.
> 
> Just going to repost your code simplified and formatted for reference:
> 
> ```js
> import gulp from 'gulp'
> import newer from 'gulp-newer'
> 
> const app = 'app'
> const dist = 'dist'
> 
> const config = {
>   app: {
>     img: app + '/images//.{jpg,jpeg,png,gif,webp}'
>   },
>   dist: {
>     img: dist + '/images//*'
>   }
> }
> export const imgtinify = () =>
>   gulp
>     .src(config.app.img, {
>       base: 'app',
>       encoding: false,
>       buffer: false
>     })
>     .pipe(newer(config.dist.img))
>     .pipe(gulp.dest(dist))
> 
> export default gulp.series(imgtinify)
> ```
> 
> Are you able to reproduce this with the above code, and does removing gulp-newer also reproduce it? Could you provide an example of a before and after file so I can look at the binary?

I checked the test assembly with the code you wrote and got the same result.

**Comment by phated** [comment]
I don't have much input right now, but I'm noticing that the globs look very broken. Perhaps this isn't even picking up the files at all and the output is an older generated file

**Comment by arturvader** [comment]
when I copied the code here some characters are not published (double *, /images/**/ )

**Comment by yocontra** [comment]
To simplify further and fix the globs in the code, does this reproduce?

```js
import gulp from 'gulp'
import newer from 'gulp-newer'

export const imgtinify = () =>
  gulp
    .src('app/images/**/*.{jpg,jpeg,png,gif,webp}', {
      base: 'app',
      encoding: false,
      buffer: false
    })
    .pipe(newer(config.dist.img))
    .pipe(gulp.dest('dist'))

export default gulp.series(imgtinify)
```

**Comment by arturvader** [comment]
> To simplify further and fix the globs in the code, does this reproduce?
> 
> ```js
> import gulp from 'gulp'
> 
> export const imgtinify = () =>
>   gulp
>     .src('app/images/**/*.{jpg,jpeg,png,gif,webp}', {
>       base: 'app',
>       encoding: false,
>       buffer: false
>     })
>     .pipe(gulp.dest('dist'))
> 
> export default gulp.series(imgtinify)
> ```
> 
> Is it possible that you ran it without encoding: false, it produced broken images, then you added encoding: false but the new images didn't get written because gulp-newer is just looking at the mtime of the original file? Try clearing the dist folder out before reproducing again.

The source file was made in Photoshop, I deleted all the files in dist before launching gulp. This is a test build, the problem was discovered in 2 new projects with a lot of image files and gulp version 5 was installed for the first time. In projects where the old version (previous) of gulp remains, everything works correctly.

**Comment by arturvader** [comment]
Please try to create a test project with the settings that I specified. Maybe I made a mistake somewhere. I tried it on my MacBook Pro and on Linux (Raspberry PI 5). Sorry for my English, it's not my native language.

**Comment by Nezarik-Intmax** [comment]
Try to add `removeBOM: false` to src options

**Comment by pepelsbey** [comment]
I have a similar issue with fonts that get corrupted during the build when using gulp@5.0.0. Browsers just won’t load them for some reason. Once I roll back to gulp@4.0.2, everything is fine.

**Comment by mjot** [comment]
I had the same problem with this simple task for copying font files:
```js
async function fontsBuild(site) {
    let path = setPathsForSite(site);
    let fontsConfig = {
        dir: path.src + 'fonts/',
        src: path.src + 'fonts/*.{woff,woff2,eot,ttf,svg}',
        build: path.dst + 'assets/fonts/'
    };

    const directoryExists = await checkDirectory(fontsConfig.dir);
    if (directoryExists) {
        return gulp.src(fontsConfig.src, { buffer: true })
            .pipe(newer(fontsConfig.build))
            .pipe(gulp.dest(fontsConfig.build));
    }
}
```
By adding `encoding: false` to gulp.src like mentioned from @yocontra its works again:
```
return gulp.src(fontsConfig.src, { buffer: true, encoding: false })
```

**Comment by arturvader** [comment]
> Try to add `removeBOM: false` to src options

I tried what you wrote, it didn't help.

**Comment by vkorytska** [comment]
This is probably related to https://github.com/gulpjs/vinyl-fs/issues/351

**Comment by DuaelFr** [comment]
I have the same issue for both my images and my fonts.
Text based files seem to be acting normal.

**Comment by BlueAbdul** [comment]
I had the same problem for pictures and fonts that was resolved by setting buffer to true and removeBOM to false.
```javascript
...
return gulp
          .src(your_path, {
            buffer: true,
            removeBOM: false
          })
...
```

**Comment by yavorski** [comment]
We have identical issue when creating `.woff2` font file from multiple svg images using [gulp-iconfont](https://www.npmjs.com/package/gulp-iconfont). 
We only updated to v5 and it stopped working, no other changes are involved. Reverting to v4 fixes the issue.

**Comment by arturvader** [comment]
> У нас возникла идентичная проблема при создании `.woff2`файла шрифта из нескольких изображений SVG с использованием [gulp-iconfont](https://www.npmjs.com/package/gulp-iconfont) . Мы обновились только до версии 5, и она перестала работать, никаких других изменений не произошло. Возврат к версии 4 решает проблему.

Many plugins stopped working. Here is a list of some that stopped working after installing gulp 5: vinyl-ftp, gulp-webp, gulp-sftp (I'm not sure about this plugin, it worked together with vinyl-ftp, I didn't sell it separately)

**Comment by rustscr1pt** [comment]
Yeah, I had the same issue today. I returned to 4.0.2 version and it works fine.

**Comment by ernilambar** [comment]
I was having same issue of corrupted image. Adding following arguments in `src()` fixed the issue.

```js
{
  buffer: true,
  removeBOM: false
}
```

Example:

```js
return gulp.src( [ 'src/images/**/*.*' ], {
  buffer: true,
  removeBOM: false
  } )
  .pipe( gulp.dest( 'assets/images' ) );
} );
```

**Comment by arturvader** [comment]
> ```js
> true
> ```

I confirm that after setting the buffer value to true (buffer: true), the file was copied without changes and working.

**Comment by BlueAbdul** [comment]
> > ```js
> > true
> > ```
> 
> I confirm that after setting the buffer value to true (buffer: true), the file was copied without changes and working.

Yeah it fixes the problem, I don't know why my comment got minimized as I'm only trying to help...

**Comment by woody-li** [comment]
> { buffer: true, removeBOM: false }

It still don't working for font sources, such as [gulp-iconfont](https://www.npmjs.com/package/gulp-iconfont).

**Comment by PRR24** [comment]
According to the doc, buffer is true by default. It changing it affects the outcome, there is possibly another smaller defect.

**Comment by PRR24** [comment]
Took a look into the code.
The specific issue is likely to be caused by invalid encoding detection in read-buffer.js (or read-stream.js) returning _utf8_, while it is expeted to return _null_ for binary images.
The reason why it returns _utf8_, not _null_, is that _resolve_ function is given a second argument _file_ that is probably expected to be used for encoding detection, but the resolve function does not have the second argument at all.

**Comment by PRR24** [comment]
After digging around a bit in the code and being initially mislead by unused arguments, I found few issues with stream writing part of the code.
* the optResolver used there is not in sync with the resolver in _gulp.src_, so the encoding used in stream writing differs from stream reading (being _utf8_ always). Feels like a minefield, especially as write_stream is using read_stream functions.
* line 34 in write_stream.js is probably meant to read _... codec.enc !== ..._ as _encoding.enc_ makes no sense.

After fixing the line 34, my scripts started to work correctly.
That said, I think the stream functionality would require some additional love.

**Comment by PRR24** [comment]
@KarelWintersky You may try, if an in-place patch of
```node_modules/vinyl-fs/lib/dest/write-contents/write-stream.js:34```
as per above solves the problem for you.
If yes, this is a global fix.

**Comment by yocontra** [comment]
> I think I found the cause of the problem: line 34 in write-stream.js is currently reading _... encoding.enc !== ..._. It should be _... codec.enc !== ..._ instead. After making the change, all my scripts started to work correctly.
> 
> While looking around in the code, two additional things caught my eye:
> 
>     * optResolver used in write-stream is not in sync with the resolver in _gulp.src_, so the encoding used in stream writing differs from stream reading. As write-stream is calling stream reading functions, it feels bit scary.
> 
>     * Resolver.resolve is called with second argument (file), but the function has only a single argument.
> 
> 
> Possibly these are there for a reason, but just wanted to point them out.

Can you submit a PR or open an issue on vinyl-fs? Thank you for looking deeper into this.

**Comment by keitetran** [comment]
> Add `{encoding: false}` to any gulp.src call where you are using binary files - see now docs and release blog for v5 for info.

I was add this for copy excel file.

**Comment by aditya305** [comment]
I've got a straightforward solution for you! Downgrading Gulp to version 4.0.2 should resolve the issue temporarily. To see this in action, you can check out my repository - [https://github.com/Brewnbeer/html_biolerplate](url)

**Comment by alex-lenk** [comment]
> encoding: false



> ```js
> {
>   buffer: true,
>   removeBOM: false
> }
> ```

Praise be to you, kind person. You have solved my problem. I was already considering switching to version 4.

**Comment by twolfson** [comment]
In `gulp.spritesmith` we had a comparable problem but the fix had to also include add a `{encoding: false}` on the `.dest()` as well

https://github.com/twolfson/gulp.spritesmith/issues/159

Example of fixed `Gulpfile.js`:

```js
var gulp = require('gulp');
var spritesmith = require('gulp.spritesmith');

gulp.task('sprite', function () {
  var spriteData = gulp.src('images/*.png', {encoding: false}).pipe(spritesmith({
    imgName: 'sprite.png',
    cssName: 'sprite.css'
  }));
  return spriteData.pipe(gulp.dest('path/to/output/', {encoding: false}));
});
```

**Comment by ImKubass** [comment]
For me none of these settings works. Once I am on gulp 5. I tried every combinations of there settings. None have work for me. I tried Node 20 and Node 23 same negative result
Version 4 works fine
``` js
const copyImages = () => {
  return src(`${build.images}**/*`, {
    buffer: true,
    removeBOM: false,
    encoding: false,
  })
    .pipe(plumber())
    .pipe(dest(`${themePath}images/`))
}
```

**Comment by GeorgeFlorian** [comment]
The following code produces broken images, after decompressing:
```
import gulp from "gulp";
import gzip from "gulp-gzip";
import gulpIf from "gulp-if";
import path from "path";

// Define file type checks
const isTextFile = (file) => /\.(html|css|js)$/.test(path.extname(file.path));
const isBinaryFile = (file) =>
  /\.(png|jpg|jpeg|gif|svg|ico)$/.test(path.extname(file.path));

export function compress() {
  return (
    gulp
      .src("dist/**/*")
      // Gzip text files with default encoding
      .pipe(
        gulpIf(
          isTextFile,
          gzip({
            gzipOptions: { level: 9 },
            append: true,
          })
        )
      )
      // Gzip binary files with { encoding: false }
      .pipe(
        gulpIf(
          isBinaryFile,
          gzip({
            gzipOptions: { level: 9 },
            append: true,
            encoding: false,
            buffer: true,
            removeBOM: false,
          })
        )
      )
      .pipe(gulp.dest("data"))
  ); // Output to LittleFS data directory
}

gulp.task("compress", compress);
```

**Comment by parachvte** [comment]
The encoding should defaults to false.

upgraded from v4, it took me a long while to recognize it's the encoding behavior changed
